packages/components/managed-data-table/src/data-table-settings/column-manager/README.md

```diff
@@ -0,0 +1,41 @@
+# Column Manager
+
+## Usage
+
+```js
+import ColumnManager from '@commercetools-local/core/components/column-manager';
+```
+
+### `<ColumnManager />`
+
+Shows a ColumnManager component to customize columns in tables. The component allows searching for available columns and order them by dragging and dropping when clicking on the Drag&Drop button which is just besides the title.
+
+#### Usage
+
+```js
+<ColumnManager
+  availableColumns={availableColumns}
+  selectedColumns={this.state.selectedColumns}
+  onUpdateColumns={this.handleUpdateColumns}
+/>
+```
+
+The column manager also receives an array `hiddenColumns` with the columns which are in the `availableColumns` array but are not included in the `selectedColumns` one.
+
+#### Props
+
+| Props              | Type              | Required | Values | Default | Description                                                                                                                           |
+| ------------------ | ----------------- | :------: | ------ | ------- | ------------------------------------------------------------------------------------------------------------------------------------- |
+| `availableColumns` | array of `object` |    ✅    | -      | -       | Sets the available columns to be selected in the column manager component.                                                            |
+| `selectedColumns`  | array of `object` |    ✅    | -      | -       | The selected columns to display in the table.                                                                                         |
+| `hiddenColumns`    | array of `object` |    ✅    | -      | -       | The hidden columns to display in the hidden section.                                                                                  |
+| `onUpdateColumns`  | `func`            |    ✅    | -      | -       | Function that is called when the selectedColumns change. This allows the developer to execute whatever is needed to update the state. |
```

Comment:
Should we have a generic `onChange` which passes up e.g. `{ selectedColumns }` in case they change? Just wondering if that's more scalable on the long run when adding more settings.

Reply:
@tdeekens @jonnybel the props `areHiddenColumnsSearchable`, `searchHiddenColumns` and `searchHiddenColumnsPlaceholder` are needed as the parent will decide when the search input for the hidden fields is visible. This means more props to set from the parent related to the column manager. 

This makes me wonder if we should use an object prop for the the _column manager_ and another one for the _display settings_ in order to avoid such larger amount of props:

```
...
displaySettingsProps: PropTypes.shape({
  useDisplaySettings: PropTypes.bool,
  isCondensed: PropTypes.bool,
  isWrappingText: requiredIf(
    PropTypes.bool,
    (props) => props.displaySettingsProps.useDisplaySettings
  ),  
  onIsCondensedChange: requiredIf(
    PropTypes.func,
    (props) => props.displaySettingsProps.useDisplaySettings
  ),
  onIsWrappingTextChange: requiredIf(
    PropTypes.func,
    (props) => props.displaySettingsProps.useDisplaySettings
  ),
}),
columnManagerProps: PropTypes.shape({
  useColumnManager: PropTypes.bool,
  visibleColumnKeys: requiredIf(
    PropTypes.array,
    (props) => props.columnManagerProps.useColumnManager
  ),
  hideableColumns: requiredIf(
    PropTypes.array,
    (props) => props.columnManagerProps.useColumnManager
  ),
  onUpdateColumns: requiredIf(
    PropTypes.func,
    (props) => props.columnManagerProps.useColumnManager
  ),
  areHiddenColumnsSearchable: PropTypes.bool,
  searchHiddenColumns: requiredIf(
    PropTypes.func,
    (props) => props.columnManagerProps.useColumnManager && props.columnManagerProps.areHiddenColumnsSearchable
  ),
  searchHiddenColumnsPlaceholder: PropTypes.string,
}),
...
```
What do you think?

Reply:
Yes, considering that the `ManagedTable` already has all the `Table` props, it can help to tidy up the props a bit.

Reply:
Hm. My main question was if we given we have multiple parts of states would want to

1. `foo: string`, `bar: object`, `baz: number`
   - Have `onChangeFoo`, `onChangeBar` and `onChangeBaz`
2. `foo: string`, `bar: object`, `baz: number`
   - Have `onChange({ foo, bar, baz })` always being called with all parts of controlled state
3. `settings: { foo: string`, `bar: object`, `baz: number }`
   - Have `onChangeSettings`

In your example the props are grouped by where they have to go inside the component. That itself can be worthwhile regardless but props shouldn't end with `*Props` but just be `columnManager` and `displaySettings`. Then inside the question above still applies. 

I also wonder what the goal of controlling the component is or where it can be uncontrolled. For this I lack a bit of context. An wild idea is to have a `stateReducer` prop over many `onChange` and `prop`-pairs. Similar to downshift. The reducer is called whenever state changes and allows the parent to control how it would be updated.

Reply:
As an additional note: nesting or grouping props can always be nice. Still, consumer will always do this `settings: { foo: 'bar' }, columnsMangaer: { bar: 'bar' }` which means new objects on each render. With flat props you get around that more easily to don't have to guard against rendering too often due to your component's API.

Reply:
> As an additional note: nesting or grouping props can always be nice. Still, consumer will always do this `settings: { foo: 'bar' }, columnsMangaer: { bar: 'bar' }` which means new objects on each render. With flat props you get around that more easily to don't have to guard against rendering too often due to your component's API.

In this case I think any change over the setting will actually cause a re-rendering of the table as you are wrapping/not wrapping the table, condensing/not condensing and changing the order or the visible columns.

Reply:
I really like the idea of the `stateReducer`. What do you think @jonnybel ? Should we go for it?

Reply:
I like that idea too! I'd say we go with it 🙂